core/src/banking_stage/forward_packet_batches_by_accounts.rs

```diff
@@ -104,39 +74,86 @@ impl ForwardPacketBatchesByAccounts {
+
+        let batch_vote_limit = block_cost_limits::MAX_VOTE_UNITS.saturating_div(limit_ratio as u64);
+        let batch_block_limit =
+            block_cost_limits::MAX_BLOCK_UNITS.saturating_div(limit_ratio as u64);
+        let batch_account_limit =
+            block_cost_limits::MAX_WRITABLE_ACCOUNT_UNITS.saturating_div(limit_ratio as u64);
+
+        let mut cost_tracker = CostTracker::default();
+        cost_tracker.set_limits(
+            batch_account_limit.saturating_mul(number_of_batches as u64),
+            batch_block_limit.saturating_mul(number_of_batches as u64),
+            batch_vote_limit.saturating_mul(number_of_batches as u64),
+        );
+        Self {
+            forward_batches,
+            cost_tracker,
+            batch_vote_limit,
+            batch_block_limit,
+            batch_account_limit,
+        }
     }
 
-    /// packets are filled into first available 'batch' that have space to fit it.
     pub fn try_add_packet(
         &mut self,
         sanitized_transaction: &SanitizedTransaction,
         immutable_packet: Arc<ImmutableDeserializedPacket>,
         feature_set: &FeatureSet,
     ) -> bool {
-        for forward_batch in self.forward_batches.iter_mut() {
-            if forward_batch
-                .try_add(sanitized_transaction, immutable_packet.clone(), feature_set)
-                .is_ok()
-            {
-                return true;
+        let tx_cost = CostModel::calculate_cost(sanitized_transaction, feature_set);
+
+        if let Ok(updated_costs) = self.cost_tracker.try_add(&tx_cost) {
+            let batch_index = self.get_batch_index_by_updated_costs(&tx_cost, &updated_costs);
+
+            if let Some(forward_batch) = self.forward_batches.get_mut(batch_index) {
+                forward_batch.forwardable_packets.push(immutable_packet);
+            } else {
+                // A successfully added tx_cost means it does not exceed block limit, nor vote
+                // limit, nor account limit. batch_index calculated as quotient from division
+                // will not be out of bounds.
+                unreachable!("batch_index out of bounds");
             }
+            true
+        } else {
+            false
         }
-        false
     }
 
     pub fn iter_batches(&self) -> impl Iterator<Item = &ForwardBatch> {
         self.forward_batches.iter()
     }
+
+    // Successfully added packet should be placed into the batch where no block/vote/account limits
+    // would be exceeded. Eg, if by block limit, it can be put into batch #1; by vote limit, it can
+    // be put into batch #2; and by account limit, it can be put into batch #3; then it should be
+    // put into batch #3 to satisfy all batch limits.
+    fn get_batch_index_by_updated_costs(
+        &self,
+        tx_cost: &TransactionCost,
+        updated_costs: &UpdatedCosts,
+    ) -> usize {
+        let batch_index_by_block_limit = updated_costs.updated_block_cost
+            / match tx_cost {
```

Comment:
no, those limits must be above zero, otherwise nothing gets into block. But in any case, will change to checked_div in a follow up PR, so to be correct 😄